lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/CatalogFunctionAccessControlRule.java

```diff
@@ -0,0 +1,105 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base.security;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.plugin.base.security.FunctionAccessControlRule.FunctionPrivilege;
+import io.trino.spi.connector.CatalogSchemaRoutineName;
+import io.trino.spi.function.FunctionKind;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+
+public class CatalogFunctionAccessControlRule
+{
+    public static final CatalogFunctionAccessControlRule ALLOW_ALL = new CatalogFunctionAccessControlRule(
+            FunctionAccessControlRule.ALLOW_ALL,
+            Optional.empty());
+
+    private final FunctionAccessControlRule functionAccessControlRule;
+    private final Optional<Pattern> catalogRegex;
+
+    @JsonCreator
+    public CatalogFunctionAccessControlRule(
+            @JsonProperty("privileges") Set<FunctionPrivilege> privileges,
+            @JsonProperty("user") Optional<Pattern> userRegex,
+            @JsonProperty("role") Optional<Pattern> roleRegex,
+            @JsonProperty("group") Optional<Pattern> groupRegex,
+            @JsonProperty("schema") Optional<Pattern> schemaRegex,
+            @JsonProperty("function_kinds") @JsonAlias("functionKinds") Set<FunctionKind> functionKinds,
+            @JsonProperty("function") Optional<Pattern> tableFunctionRegex,
+            @JsonProperty("catalog") Optional<Pattern> catalogRegex)
+    {
+        this(new FunctionAccessControlRule(privileges, userRegex, roleRegex, groupRegex, schemaRegex, functionKinds, tableFunctionRegex), catalogRegex);
+    }
+
+    private CatalogFunctionAccessControlRule(FunctionAccessControlRule functionAccessControlRule, Optional<Pattern> catalogRegex)
+    {
+        this.functionAccessControlRule = requireNonNull(functionAccessControlRule, "tableFunctionAccessControlRule is null");
+        this.catalogRegex = requireNonNull(catalogRegex, "catalogRegex is null");
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, String functionName)
+    {
+        return functionAccessControlRule.matches(user, roles, groups, functionName);
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, FunctionKind functionKind, CatalogSchemaRoutineName functionName)
+    {
+        if (!catalogRegex.map(regex -> regex.matcher(functionName.getCatalogName()).matches()).orElse(true)) {
+            return false;
+        }
```

Comment:
nit: wanted understand that and later to optimise, but looks it's already the best representation. 

Maybe something like extract constant `MATCH_ON_EMPTY = true` and `.orElse(true)` -> `.orElse(MATCH_ON_EMPTY)` to make code more self describable. But maybe it's already obvious.